src/vm/diagnosticprotocolhelper.cpp

```diff
@@ -0,0 +1,81 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "fastserializer.h"
+#include "diagnosticprotocolhelper.h"
+#include "diagnosticsipc.h"
+#include "diagnosticsprotocol.h"
+
+#ifdef FEATURE_PERFTRACING
+
+static void WriteStatus(uint64_t result, IpcStream* pStream)
+{
+    uint32_t nBytesWritten = 0;
+    bool fSuccess = pStream->Write(&result, sizeof(result), nBytesWritten);
+    if (fSuccess)
+    {
+        fSuccess = pStream->Flush();
+    }
+}
+
+void DiagnosticProtocolHelper::GenerateCoreDump(IpcStream* pStream)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(pStream != nullptr);
+    }
+    CONTRACTL_END;
+
+    HRESULT hr = S_OK;
+
+    if (pStream == nullptr)
+        return;
+
+    // TODO: Read within a loop.
+    uint8_t buffer[IpcStreamReadBufferSize] { };
+    uint32_t nNumberOfBytesRead = 0;
+    bool fSuccess = pStream->Read(buffer, sizeof(buffer), nNumberOfBytesRead);
```

Comment:
Instead of having an arbitrary number of bytes that we attempt to read from the pipe, could you calculate the total size of the payload on the client and send it first?  This way you only ever need to create a buffer that is exactly the right size and the call to `Read` won't be asked to read 8 KB.  This will be a non-issue once the IPC protocol work is finished, so I guess this isn't a big deal if this works for now.

Reply:
I was lazy and just copied the code from the eventpipeprocotol.cpp.  I also wanted to use a holder (ArrayHolder or NewArrayHolder) and not put on the stack.

If the protocol is going to change soon anyway, I'm going to leave it alone.  I assume that part of the new protocol will be the necessary payload size.